src/sources/stdin.rs

```diff
@@ -75,17 +81,33 @@ where
         .host_key
         .unwrap_or_else(|| log_schema().host_key().to_string());
     let hostname = crate::get_hostname().ok();
+    let decoder = config.decoding.build()?;
 
     let (mut sender, receiver) = mpsc::channel(1024);
 
-    // Start the background thread
+    // Spawn background thread with blocking I/O to process stdin.
+    //
+    // This is recommended by Tokio, as otherwise the process will not shut down
+    // until another newline is entered. See
+    // https://github.com/tokio-rs/tokio/blob/a73428252b08bf1436f12e76287acbc4600ca0e5/tokio/src/io/stdin.rs#L33-L42
     thread::spawn(move || {
         info!("Capturing STDIN.");
 
-        for line in stdin.lines() {
-            if executor::block_on(sender.send(line)).is_err() {
-                // receiver has closed so we should shutdown
-                return;
+        loop {
+            let (buffer, len) = match stdin.fill_buf() {
+                Ok(buffer) if buffer.is_empty() => break, // EOF.
+                Ok(buffer) => (Ok(Bytes::copy_from_slice(buffer)), Some(buffer.len())),
+                Err(error) if error.kind() == std::io::ErrorKind::Interrupted => continue,
+                Err(error) => (Err(error), None),
+            };
+
+            if let Some(len) = len {
+                stdin.consume(len);
+            }
```

Comment:
If `stdin.consume(0)` is a no-op, this could be simplified by dropping the `Option` around `len` and using `0` for the error case.

Reply:
Changed in 0d2a57b456febf1aef65bc6329572718dae2ba2c